### DISCUSSION TITLE
--mount=type=cache location

### DISCUSSION THREAD

**Discussion by tmds** [containers/podman#15612] [Q&A] [answered]
Where do the files get stored when using `--mount=type=cache,id=<id>,target=<target>,Z`?
And where when leaving out the `<id>`?

cc @mheon @rhatdan

**Comment by flouthoc** [comment, marked as answer]
Hi @tmds, It gets stored on your host `$TMPDIR/buildah-cache` and if `$TMPDIR` is not set it is `/var/tmp/buildah-cache` and when `id` option is selected it becomes `$TMPDIR/buildah-cache/<id>` or `/var/tmp/buildah-cache/<id>` depending upon if `$TMPDIR` was set.

I hope this answers your question. 

Ref: https://github.com/containers/buildah/blob/main/pkg/parse/parse.go#L49

**Reply by tmds** [reply]
Thanks!

**Reply by rhatdan** [reply]
We should be checking to make sure that the content in /var/tmp/buildah-cache/<id> is created by the <id> and/or make sure that /var/tmp/buildah-cache/<id> has permissions 0700

**Reply by marwatk** [reply]
Doesn't this mean that only a single non-root user can use buildah cache on a system? I would think the cache would need to be namespaced per-user?

**Reply by marwatk** [reply]
Followup, is there a way to flush this cache using `podman`? It's left around after a `podman system reset`.

**Reply by eschnett** [reply]
I'm using podman on macOS. Apparently, the "host" there is the VM in which the podman containers run. To see the cache, you need to `podman machine ssh`, and then `ls /var/tmp`.

**Reply by flouthoc** [reply]
@marwatk Recently added `buildah prune` clears the build cache https://github.com/containers/buildah/pull/4490, soon podman's prune will do it as well.

**Reply by sanmai-NL** [reply]
@flouthoc How does this work with builds from the Podman container image? `TMPDIR`isn't set.

```console
# nerdctl run --privileged --rm -it quay.io/podman/stable:latest /bin/sh
sh-5.2# vi Containerfile
sh-5.2# podman build .
STEP 1/2: FROM alpine
Resolved "alpine" as an alias (/etc/containers/registries.conf.d/000-shortnames.conf)
Trying to pull docker.io/library/alpine:latest...
Getting image source signatures
Copying blob 9fda8d8052c6 done  
Copying config f6648c04cd done  
Writing manifest to image destination
STEP 2/2: RUN --mount=type=cache,target=/srv/dist,z touch blabla
COMMIT
--> 6d411b713297
6d411b713297990b68a84d2786baa2565301758ad8cc010ba2f2c66ec75b6cf3
sh-5.2# find / -type f -iname 'blabla'
/var/lib/containers/storage/overlay/8ec8d4309baa3c18ae89deedcd948303d4faddfcf6dd5b440c8255b319fc673d/diff/blabla
sh-5.2# printenv | sort
DISTTAG=f38container
FGC=f38
HOME=/root
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PWD=/
SHLVL=1
TERM=xterm
_=/usr/bin/printenv
_CONTAINERS_USERNS_CONFIGURED=
container=oci
sh-5.2# 
```

**Reply by flouthoc** [reply]
@sanmai-NL have you checked `/var/tmp/buildah-cache` , cache should be there if `TMPDIR` is not set.

**Reply by cgwalters** [reply]
> @marwatk Recently added buildah prune clears the build cache containers/buildah#4490, soon podman's prune will do it as well.

But we don't ship `buildah` in the default podman-machine-os...any thoughts on exposing something to prune this via `podman build...` maybe?

**Reply by flouthoc** [reply]
> > @marwatk Recently added buildah prune clears the build cache [containers/buildah#4490](https://github.com/containers/buildah/pull/4490), soon podman's prune will do it as well.
> 
> But we don't ship `buildah` in the default podman-machine-os...any thoughts on exposing something to prune this via `podman build...` maybe?

@cgwalters Did you mean `podman prune ..` or `podman build ..` ?   Could you suggest a bit more about the cli of `podman build ..` implementing this I was thinking making this part of `podman prune ...`, However I thought code `builldah prune` and `podman prune` is similar, but I'll check and confirm again.

**Reply by cgwalters** [reply]
There's no `podman prune`, did you mean `podman image prune`?

I did find it slightly surprising that build caches were not pruned with `podman image prune`, and perhaps we just take it as a bug and do it there? Dunno.

**Reply by flouthoc** [reply]
Ah i meant `podman image prune` i'll check and open a PR against `image prune` it if it doesn't already cleans build cache. I also wonder if this fits more in `podman build prune .. ` ? A new command which can remove build containers and build cache but  I am not sure if everybody would encourage creating a new command just for this.

Thanks for reporting this.

**Reply by flouthoc** [reply]
@cgwalters Created a PR https://github.com/containers/podman/pull/23846

**Reply by parkan** [reply]
did this behavior change recently? I am running:

```
RUN --mount=type=cache,id=pip,target=$HOME/.cache/pip pip3 install --upgrade pip wheel setuptools ...
```
with no TMPDIR set and the only directories with similar names are:

```
❯ ls -lh /var/tmp/buildah-cache-1000/
Permissions Size User    Date Modified Name
drwxr-xr-x@    - arkadiy  9 Jan 13:47  41cddf472639a12b
drwx------@    - arkadiy  9 Jan 14:57  buildah-cache-lockfiles
drwxr-xr-x@    - arkadiy  9 Jan 14:57  fed5038bf2f427ab
```

furthermore the cache does not seem to actually be working (subsequent runs re-download the pip packages and all subdirs in the buildah-cache-1000 dir are empty

(my uid is 1000 and podman version is 5.3.1)

**Reply by nalind** [reply]
Yes, https://github.com/containers/buildah/pull/5778 changed it so that we don't directly use the "id" (or "target", if "id" isn't specified) in constructing the path to the top level of the cache - whichever one we're given is hashed to produce a subdirectory name to avoid shenanigans.

**Reply by nalind** [reply]
When I tried your example instruction using registry.redhat.io/ubi9/python-312 as a base image, I had to specify the "uid" option to ensure that the cache was writeable by the non-root user the image used as a default, and had to use pip3's "--cache-dir" option to actually use the cache directory, but that could equally be explained by my not being familiar with pip.

**Reply by parkan** [reply]
@nalind thank you for the clarification on the paths, the uid issue may be silently affecting me as well, will check this

pip should use $HOME/.cache/pip as the cache dir by default unless XDG_CACHE_HOME is set, which may have been the case in your trial

**Reply by nalind** [reply]
registry.redhat.io/ubi9/python-312 does not set $XDG_CACHE_HOME, but it does set $HOME.  Turns out that image is meant for use as a base image for a source-to-image build, so its default command is not an interactive shell, meaning I probably should have been using a different base image, since my quick test was not a source-to-image build.
